### PR TITLE
stats: More StatName conversions

### PR DESCRIPTION
**Pull request by jmarantz**
Description: Avoid using scope.counter(string) in Zookeeper and ExtAuthz, factoring out some infrastructure for looking up dynamically constructed stat-names for shared use. 6 more files to go and then #4980 can hopefuly be merged, addressing #4196.
Risk Level: low
Testing: //test/...
Docs Changes: n/a
Release Notes: n/a

### CODE REVIEW

**Comment by jmarantz** [comment]
@lizan I had previously thought this PR was all green; but now it's in this "waiting for status..." state. How do I trigger a re-test for this flavor of CI?

File: source/extensions/filters/http/dynamo/dynamo_stats.cc [hunk 1]
```diff
@@ -16,44 +15,44 @@ namespace HttpFilters {
...
+  RequestParser::forEachStatString(
+      [this](const std::string& str) { stat_name_set_.rememberBuiltin(str); });
 }
 
-Stats::SymbolTable::StoragePtr DynamoStats::addPrefix(const std::vector<Stats::StatName>& names) {
-  std::vector<Stats::StatName> names_with_prefix{prefix_};
-  names_with_prefix.reserve(names.end() - names.begin());
+Stats::SymbolTable::StoragePtr DynamoStats::addPrefix(const Stats::StatNameVec& names) {
+  Stats::StatNameVec names_with_prefix;
+  names_with_prefix.reserve(1 + names.end() - names.begin());
```
Comment (snowp) [inline source/extensions/filters/http/dynamo/dynamo_stats.cc]: maybe do `1 + names.size()` instead?

File: source/extensions/filters/network/zookeeper_proxy/filter.cc [hunk 2]
```diff
@@ -18,8 +18,18 @@ namespace ZooKeeperProxy {
 
 ZooKeeperFilterConfig::ZooKeeperFilterConfig(const std::string& stat_prefix,
                                              const uint32_t max_packet_bytes, Stats::Scope& scope)
-    : scope_(scope), max_packet_bytes_(max_packet_bytes), stat_prefix_(stat_prefix),
-      stats_(generateStats(stat_prefix, scope)) {}
+    : scope_(scope), max_packet_bytes_(max_packet_bytes), stats_(generateStats(stat_prefix, scope)),
+      stat_name_set_(scope.symbolTable()), stat_prefix_(stat_name_set_.add(stat_prefix)),
+      auth_(stat_name_set_.add("auth")) {
+  // https://zookeeper.apache.org/doc/r3.1.2/zookeeperProgrammers.html#sc_BuiltinACLSchemes
+  // lists commons schemes: "world", "auth", "digest", "host", and "ip". These are used in
```
Comment (snowp) [inline source/extensions/filters/network/zookeeper_proxy/filter.cc]: @rgs1 does this match your understanding of common stat names?

File: source/common/stats/symbol_table_impl.h [hunk 3]
```diff
@@ -630,5 +630,44 @@ class StatNameStorageSet {
...
+// can be mutated at runtime.
+class StatNameSet {
+public:
+  explicit StatNameSet(SymbolTable& symbol_table);
+
+  /**
+   * Adds a string to the builtin map, which is not mutex protected. This
+   * map is always consulted first as a hit there means no lock is required.
+   */
+  void rememberBuiltin(absl::string_view str);
```
Comment (snowp) [inline source/common/stats/symbol_table_impl.h]: Maybe reiterate the bit about "should be called during initialization" for the comment here? Might also be nice to have a note what exactly initialization means: is it just before any stats are accessed?

File: source/common/stats/symbol_table_impl.cc [hunk 4]
```diff
@@ -429,5 +429,35 @@ void StatNameList::clear(SymbolTable& symbol_table) {
...
+    absl::MutexLock lock(&mutex_);
+    stat_name = pool_.add(str);
+  }
+  builtin_stat_names_[str] = stat_name;
+}
+
+Stats::StatName StatNameSet::getStatName(absl::string_view token) {
+  // If token was recorded as a built-in during initialization, we can
+  // service this request lock-free.
+  auto iter = builtin_stat_names_.find(token);
```
Comment (rgs1) [inline source/common/stats/symbol_table_impl.cc]: nit: `const auto` ?

File: source/extensions/filters/network/zookeeper_proxy/filter.cc (hunk 2 above)
Comment (rgs1) [reply source/extensions/filters/network/zookeeper_proxy/filter.cc]: Mostly correct, but lets reference a more updated link instead:

https://zookeeper.apache.org/doc/r3.5.4-beta/zookeeperProgrammers.html#sc_BuiltinACLSchemes

And include the `x509` scheme as well. 

[or, i can do it in a follow-up]

Comment (jmarantz) [reply source/extensions/filters/network/zookeeper_proxy/filter.cc]: good catch, thanks.

**Comment by jmarantz** [comment]
Any other comments? This is blocking the reviewability of https://github.com/envoyproxy/envoy/pull/7842 which I think is the last thing we need prior to reviewing #4980 and thus resolving #4196 .

**Comment by jmarantz** [comment]
Thanks @snowp -- @envoyproxy/senior-maintainers can I get another round of approval?

**Pull request merged**